common/djangoapps/student/tests/tests.py

```diff
@@ -930,3 +933,182 @@ def test_roundtrip_with_unicode_course_id(self):
         real_user = user_by_anonymous_id(anonymous_id)
         self.assertEqual(self.user, real_user)
         self.assertEqual(anonymous_id, anonymous_id_for_user(self.user, course2.id, save=False))
+
+
+@ddt.ddt
+class DashboardTestXSeriesPrograms(ModuleStoreTestCase):
+    """
+    Tests for dashboard for xseries program courses. Enroll student into
+    three programs and then try different combinations to see xseries upsell
+    messages are appearing.
+    """
+    def setUp(self):
+        super(DashboardTestXSeriesPrograms, self).setUp()
+
+        self.user = UserFactory.create(username="jack", email="jack@fake.edx.org", password='test')
+        self.course = CourseFactory.create()
+        self.course_two = CourseFactory.create()
+        self.course_three = CourseFactory.create()
+
+        CourseModeFactory.create(
+            course_id=self.course.id,
+            mode_slug='verified',
+            mode_display_name='Verified',
+            expiration_datetime=datetime.now(pytz.UTC) + timedelta(days=1)
+        )
+        self.client = Client()
+        cache.clear()
+
+    def _create_program_config(self, **kwargs):
+        """
+        DRY helper.  Create a new ProgramsApiConfig with self.DEFAULTS, updated
+        with any kwarg overrides.
+        """
+        data = dict(
+            internal_service_url="http://internal/",
+            public_service_url="http://public/",
+            api_version_number=1
+        )
+        ProgramsApiConfig(**dict(data, **kwargs)).save()
+
+    def _create_program_data(self, data):
+        """Dry method to create testing programs data."""
+        programs = {}
+        for course in data:
+            programs[unicode(course[0])] = {
+                'category': 'xseries',
+                'organization': {'display_name': 'Test Organization 1', 'key': 'edX'},
+                'marketing_slug': 'fake-marketing-slug-xseries-1',
+                'status': course[1],
+                'course_codes': [
+                    {
+                        'display_name': 'Demo XSeries Program 1',
+                        'key': 'TEST_A',
+                        'run_modes': [{'sku': '', 'mode_slug': 'ABC_1', 'course_key': 'edX/DemoX_1/Run_1'}]
```

Comment:
it seems like the course_key you generate here needs to match the course_key of the course passed in.